src/plugins/shell/shell_impl.cpp

```diff
@@ -0,0 +1,192 @@
+#include "shell_impl.h"
+#include "system.h"
+#include "global_include.h"
+#include "px4_custom_mode.h"
+#include <cmath>
+#include <functional>
+#include <array>
+#include "log.h"
+
+namespace mavsdk {
+
+void ShellImpl::init()
+{
+    using namespace std::placeholders; // for `_1`
+    _parent->register_mavlink_message_handler(
+        MAVLINK_MSG_ID_SERIAL_CONTROL,
+        std::bind(&ShellImpl::process_shell_message, this, _1),
+        this);
+}
+
+void ShellImpl::deinit() {}
+
+void ShellImpl::enable() {}
+
+void ShellImpl::disable() {}
+
+ShellImpl::ShellImpl(System& system) : PluginImplBase(system)
+{
+    _parent->register_plugin(this);
+}
+
+ShellImpl::~ShellImpl()
+{
+    _parent->unregister_plugin(this);
+}
+
+void ShellImpl::shell_command_response_async(Shell::result_callback_t& callback)
+{
+    std::lock_guard<std::mutex> lock_subscription(_result_subscription_mutex);
+    _result_subscription = callback;
+}
+
+void ShellImpl::shell_command(const Shell::ShellMessage& shell_message)
+{
+    std::lock_guard<std::mutex> lock_subscription(_result_subscription_mutex);
+    if (!_parent->is_connected()) {
+        if (_result_subscription) {
+            auto callback = _result_subscription;
+            Shell::Result arg{Shell::Result::ResultCode::NO_SYSTEM, {}};
+            _parent->call_user_callback([callback, arg]() { callback(arg); });
+            return;
+        }
+    }
+
+    {
+        std::lock_guard<std::mutex> lock(_shell_message_mutex);
+        _shell_message = shell_message;
+    }
+
+    if (!send_shell_message_mavlink()) {
+        if (_result_subscription) {
+            auto callback = _result_subscription;
+            Shell::Result arg{Shell::Result::ResultCode::CONNECTION_ERROR, {}};
+            _parent->call_user_callback([callback, arg]() { callback(arg); });
+            return;
+        }
+    }
+    if (_result_subscription) {
+        if (!shell_message.need_response) {
+            auto callback = _result_subscription;
+            Shell::Result arg{Shell::Result::ResultCode::SUCCESS, {}};
+            _parent->call_user_callback([callback, arg]() { callback(arg); });
+        } else {
+            std::lock_guard<std::mutex> lock_cookie(_shell_message_timeout_cookie_mutex);
+            if (_shell_message_timeout_cookie) {
+                _parent->unregister_timeout_handler(_shell_message_timeout_cookie);
+                _shell_message_timeout_cookie = nullptr;
+            }
+            _parent->register_timeout_handler(
+                std::bind(&ShellImpl::receive_shell_message_timeout, this),
+                shell_message.timeout / 1000.0,
+                &_shell_message_timeout_cookie);
+        }
+    }
+}
+
+void ShellImpl::receive_shell_message_timeout()
+{
+    std::lock_guard<std::mutex> lock_result(_result_mutex);
+    std::lock_guard<std::mutex> lock_subscription(_result_subscription_mutex);
+    std::lock_guard<std::mutex> lock_cookie(_shell_message_timeout_cookie_mutex);
```

Comment:
It's quite dangerous to acquire multiple locks in this way. If you try to acquire them in another order in another method you have a deadlock. I think easiest is to only have one lock for all member variables together, or alternatively you only ever acquire one lock at a time.

Reply:
I`m tried to implement some `promise`-`future` scheme.
It is not ideal I think, but it worked ok.

Reply:
We can leave it for now and try it for a bit!